### ISSUE TITLE
Dynamic Int64 equality check fails for experimental JSON Column unless Cast is used

### ISSUE THREAD

**Issue by jmwilkinson** [ClickHouse/ClickHouse#73588]
Clickhouse Version: 24.8.7.41

When using the Clickhouse JSON column type (which I believe has just moved into beta), ints are not queryable in the same way as strings or booleans: to filter against int values, they must be cast.

**How to reproduce**

Following the docs here: https://clickhouse.com/docs/en/sql-reference/data-types/newjson#alter-modify-column-to-json-type

Create a table with a JSON column, insert values, and check that the data is present:
```
CREATE TABLE test (json JSON) ENGINE = Memory;
INSERT INTO test VALUES ('{"a" : {"b" : 42}, "c" : [1, 2, 3]}'), ('{"f" : "Hello, World!"}'), ('{"a" : {"b" : 43, "e" : 10}, "c" : [4, 5, 6]}');
SELECT json FROM test;
```

Now attempt to filter by the int key `a.b`:

```
SELECT * FROM test WHERE json.a.b = 42;

SELECT *
FROM test
WHERE json.a.b = 42

Query id: 56bc4a6b-30a9-46d8-9ac3-b4b421400680

Ok.

0 rows in set. Elapsed: 0.002 sec.
```

No results.

Check the column type:
```
SELECT DISTINCT dynamicType(json.a.b) as types FROM test
```

Returns `Int64` and `None`.

Cast the column when querying, to its own dynamic type:

```
SELECT * FROM test WHERE CAST(json.a.b AS Int64) = 42;

SELECT *
FROM test
WHERE CAST(json.a.b, 'Int64') = 42

Query id: ae4e8518-8c55-436f-a7ed-f05ae15a87f2

   ┌─json───────────────────────────────┐
1. │ {"a":{"b":"42"},"c":["1","2","3"]} │
   └────────────────────────────────────┘

1 row in set. Elapsed: 0.002 sec.
```

And the data is returned as expected.

**Expected behavior**

Equality filtering should work for ints the same as strings or boolean values.

**Comment by Algunenano** [comment]
Works fine in the latest release. It was experimental in 24.8:

```
:) CREATE TABLE test (json JSON) ENGINE = Memory;


CREATE TABLE test
(
    `json` JSON
)
ENGINE = Memory

Query id: 5a13fbc4-84c7-462b-8a88-7db097480952

Ok.

0 rows in set. Elapsed: 0.000 sec. 

:) INSERT INTO test VALUES ('{"a" : {"b" : 42}, "c" : [1, 2, 3]}'), ('{"f" : "Hello, World!"}'), ('{"a" : {"b" : 43, "e" : 10}, "c" : [4, 5, 6]}');


INSERT INTO test FORMAT Values

Query id: fd642768-67b4-4018-80ec-12d397732503

Ok.

3 rows in set. Elapsed: 0.001 sec. 

:) SELECT json FROM test;


SELECT json
FROM test

Query id: c4666b9d-4f48-4aba-8611-4fb5ca82fc78

   ┌─json────────────────────────────────────────┐
1. │ {"a":{"b":"42"},"c":["1","2","3"]}          │
2. │ {"f":"Hello, World!"}                       │
3. │ {"a":{"b":"43","e":"10"},"c":["4","5","6"]} │
   └─────────────────────────────────────────────┘

3 rows in set. Elapsed: 0.001 sec. 

:) SELECT * FROM test WHERE json.a.b = 42;


SELECT *
FROM test
WHERE json.a.b = 42

Query id: 363447fa-11c5-4f6d-92f7-05ecaa50ab4a

   ┌─json───────────────────────────────┐
1. │ {"a":{"b":"42"},"c":["1","2","3"]} │
   └────────────────────────────────────┘

1 row in set. Elapsed: 0.001 sec. 

```

**Comment by jmwilkinson** [comment]
Thank you for the rapid response. I've validated this was fixed around version 24.10.

The addition of the `JSON` type is great and its encouraging to see the improvements and stability as it moves forward.
